lib/offline.js

```diff
@@ -63,52 +70,111 @@ module.exports = function(config) {
+        resources: filesAndHashes,
+        importScripts: importScripts,
+      };
 
-          if (stat.isFile() && stat.size > 2 * 1024 * 1024 && staticFileGlobs.indexOf(file) === -1) {
-            console.log(chalk.yellow.bold(file + ' is bigger than 2 MiB. Are you sure you want to cache it? To suppress this warning, explicitly include the file in the fileGlobs list.'));
-          }
-        });
+      var stream = gulp.src([__dirname + '/../templates/app/offline-worker.js'])
+        .pipe(template(replacements))
+        .pipe(conflict(rootDir))
+        .pipe(gulp.dest(rootDir));
+
+      return new Promise(function (resolve, reject) {
+        stream.on('finish', function () { resolve(); });
+        stream.on('error', function (e) { reject(e); });
       });
+    }).then(resolve, reject);
+
+    function flatGlobs(fileGlobs) {
+      return Object.keys(fileGlobs.reduce(function (matches, fileGlob) {
+        fileGlob = fileGlob.replace(path.sep, '/');
+        glob.sync(fileGlob, { nodir: true }).forEach(function (filepath) {
+          matches[filepath] = filepath;
+        });
+        return matches;
+      }, {}));
+    }
 
-      var importScripts = config.importScripts || [];
-      importScripts.forEach(function(script) {
-        var stat;
-        try {
-          stat = fs.statSync(path.join(rootDir, script));
-        } catch (ex) {
-          console.log(chalk.red.bold(script + ' doesn\'t exist.'));
-          throw ex;
-        }
-
-        if (!stat.isFile()) {
-          console.log(chalk.red.bold(script + ' is not a file.'));
-          throw new Error(script + ' is not a file.');
-        }
+    function getFilesAndHashes(files, rootDir, sizeWhiteList) {
+      var totalSize = 0;
+      var filesAndHashes = files.map(function (filepath) {
+        totalSize += countFileSize(filepath, sizeWhiteList);
+        var data = fs.readFileSync(filepath);
+        var hash = getHash(data);
+        logOk(filepath);
+        return {
+          path: filepath.replace(rootDir, function (match, offset) {
+            //XXX: The root must be the worker's directory
+            return offset === 0 ? './' : match;
+          }),
+          hash: hash,
+        };
       });
 
-      return swPrecache.write(path.join(rootDir, 'offline-worker.js'), {
-        staticFileGlobs: staticFileGlobs,
-        stripPrefix: rootDir,
-        verbose: true,
-        logger: function(message) {
-          if (message.indexOf('Caching') === 0) {
-            message = chalk.bold.green('✓ ') + message.replace('static resource ', '');
-          }
-
-          console.log(message);
-        },
-        importScripts: importScripts,
-        cacheId: cacheId,
-        ignoreUrlParametersMatching: config.ignoreUrlParametersMatching || [/./],
-        maximumFileSizeToCacheInBytes: Infinity,
+      console.log('Total precache size is about ' + prettyBytes(totalSize) +
+                  ' for ' + files.length + ' resources.');
```

Comment:
While we're here, let's update this message to make it a bit clearer. I suggest:

Total cache size is [TOTAL SIZE] for [NUMBER OF FILES] files.

Reasoning:
* precache -> cache (I think I understand what sw-precache docs mean when they say that the service worker "precaches" resources, but I'm not sure it's useful to make the distinction, especially without defining it further. And the Oghliner docs say that the service worker "caches" the files. Plus the actual API the Service Worker uses is the Cache API, and this script says it is "Caching" each file. So "cache" is the better word to use here.)
* ~~about~~ (Sure, a "pretty" number isn't accurate to an exact number of bytes, but it's still accurate enough, so "about" feels like unnecessary complexity.)
* resources -> files (Oghliner docs talk about assets and files, while sw-precache docs talk about resources and files. In both cases, the things being cached are files, and assets/resources is a more generic euphemism. Let's cut to the chase.)